tsl/src/remote/connection.c

```diff
@@ -1129,6 +1140,23 @@ set_ssl_options(const char *user_name, const char **keywords, const char **value
 	*option_start = option_pos;
 }
 
+/*
+ * Finish the connection and, optionally, save the connection error.
+ */
+static void
+finish_connection(PGconn *conn, char **errmsg)
+{
+	if (NULL != errmsg)
+	{
+		if (NULL == conn)
+			*errmsg = "invalid connection";
+		else
+			*errmsg = pchomp(PQerrorMessage(conn));
+	}
+
+	PQfinish(conn);
+}
```

Comment:
Wouldn't the code be easier to follow if this just raised an error if there were a connection problem (using `ereport`). Then it is not necessary to catch the error message and print it out, while anybody interested in catching the error can do so using `PG_TRY`.

With this approach we have different ways of handling errors: I think it would be better to stick to a single way.

Reply:
There's code that rely on this to not throw errors. Note also that throwing errors with libpq objects involved isn't always a good idea since those objects are malloc:ed and not managed through PostgreSQL's memory contexts.

Reply:
I don't understand.

You call `finish_connection` from inside `remote_connection_open_with_options_nothrow` and in each location where you call `remote_connection_open_with_options_nothrow` you immediately after check if it returns a null pointer and then you call `ereport`. If an `ereport` call inside `finish_connection` can break something, then it looks like the call immediately after `remote_connection_open_with_options_nothrow` would also break something. The only difference I see is that you pass the same memory upwards (in the `errmsg` variable), and then use it higher up in the call stack.

What am I missing?